[lldb][progress] Add progress manager class

chelcassanova opened the pull request:
Per discussions from https://github.com/llvm/llvm-project/pull/81026, it was decided that having a class that manages a map of progress reports would be beneficial in order to categorize them. This class is a part of the overall `Progress` class and utilizes a map that keeps a count of how many times a progress report category has been sent. This would be used with the new debugger broadcast bit added in https://github.com/llvm/llvm-project/pull/81169 so that a user listening with that bit will receive grouped progress reports.

chelcassanova committed:
[lldb][progress] Add progress manager class

Per discussions from https://github.com/llvm/llvm-project/pull/81026,
it was decided that having a class that manages a map of progress
reports would be beneficial in order to categorize them. This class is a
part of the overall `Progress` class and utilizes a map that keeps a
count of how many times a progress report category has been sent. This
would be used with the new debugger broadcast bit added in
https://github.com/llvm/llvm-project/pull/81169 so that a user listening
with that bit will receive grouped progress reports.

clayborg commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -119,6 +120,32 @@ class Progress {
+  // Manage the class instance internally using a std::optional
+  static std::optional<ProgressManager> &InstanceImpl();
+
+  llvm::StringMap<uint64_t> m_progress_map;
```
maybe `m_category_map`?

chelcassanova replied:
We could do `m_progress_category_map` or `m_progress_report_map` to better convey that we're using this to keep track of progress reports by category.

clayborg commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -119,6 +120,32 @@ class Progress {
+  static void Initialize();
+  static void Terminate();
+  // Control the refcount of the progress report category as needed
+  void Increment(std::string);
+  void Decrement(std::string);
```
```
void Increment(std::string category);
void Decrement(std::string category);
```

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +66,47 @@ void Progress::ReportProgress() {
   }
 }
+
+void ProgressManager::Initialize() {
+  lldbassert(!InstanceImpl() && "A progress report manager already exists.");
+  InstanceImpl().emplace();
+}
+
+void ProgressManager::Terminate() {
+  lldbassert(InstanceImpl() &&
+             "A progress report manager has already been terminated.");
+  InstanceImpl().reset();
+}
```
We can avoid the initialize/terminate if we just want to create this object on the first call to `ProgressManager::InstanceImpl()`. More comments below as to why

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +66,47 @@ void Progress::ReportProgress() {
+}
+
+std::optional<ProgressManager> &ProgressManager::InstanceImpl() {
+  static std::optional<ProgressManager> g_progress_manager;
+  return g_progress_manager;
```
The global destructor chain will still cause this to be destroyed when the main thread exits. Which is ok if we return `std::optional<ProgressManager>` from `ProgressManager::Instance()`. More below in `ProgressManager &ProgressManager::Instance()` comments.

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +66,47 @@ void Progress::ReportProgress() {
+
+  // If pair.first is not empty after insertion it means that that
+  // category was entered for the first time and should not be incremented
+  if (!pair.first->first().empty())
```
This returns a `std::pair<iterator, bool>` where the bool designates if the instertion took place. So this should be:
```
if (!pair.second)
```

JDevlieghere commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +66,47 @@ void Progress::ReportProgress() {
+
+ProgressManager::~ProgressManager() {}
+
+ProgressManager &ProgressManager::Instance() { return *InstanceImpl(); }
```
We have other subsystems that work exactly like what Chelsea did here, such as the Diagnostics and the FileSystem. They rely on the requirement that you need to call `:: Initialize ` before doing anything with it.

clayborg replied:
> We have other subsystems that work exactly like what Chelsea did here, such as the Diagnostics and the FileSystem. They rely on the requirement that you need to call `::Initialize ` before doing anything with it. It's true that this will crash if you forget to `Initialize`/`Terminate` but that's true for the `Debugger` too. This is an establishes pattern and has the advantage that it can be tested in unittest by initializing and terminating the subsystem between tests.

This is fine as long as you always check the optional then and we aren't doing that. So this can and will crash LLDB if extra threads call these APIs which is never good. So the API, in order to be safe, should be:
```
std::optional<ProgressManager &> ProgressManager::InstanceImpl() 
```
And then every user needs to check this optional. Otherwise, this:
```
ProgressManager &ProgressManager::Instance() { return *InstanceImpl(); }
```
Will crash if Terminate has been called right? the `std:optional::operator *`docs say "The behavior is undefined if *this does not contain a value."

clayborg replied:
This can cause shut down crashes in LLDB where extra threads call a progress API.

clayborg replied:
Another reason for using the `std::once` is if no one tries to listen to the eBroadcastBitProgressCategory, then nothing will need to be created. If not we are always making an instance. Granted this object isn't very expensive to create, but it is extra work that might not be needed.

clayborg replied:
I also want to say that I don't mind the need to call `ProgressManager::Initialize()` before using parts of this class. I am mostly looking out for shut down crashes of the the LLDB library where the main thread exits first while other threads are not all terminated, which does happen. All of the places we have the comment:
```
// NOTE: intentional leak to avoid issues with C++ destructor chain
```
Was from tracking down shut down crashers in LLDB.

chelcassanova replied:
To my understanding this behaviour of the Debugger caused an issue in the unit test suite when I added the test for progress reporting since it tried to initialize the debugger after another test had already done so. Could we just place the `once_flag` in `InstanceImpl()` but keep `Initialize/Terminate` to keep the pattern we already have without having the user always check the optional?

JDevlieghere replied:
Yeah, I understand why we might want to leak the vector and do in other places. But do you think we should just do that unconditionally, no matter whether it's a problem or not? The other subsystems I mention use the same approach and they don't crash in production even though they don't check the optional before dereferencing it (treating it as a precondition). 

I guess the point I'm trying to make is let's not leak it proactively, but use the stricter approach to force a crash/assert in debug builds and tests, and try to avoid these issues, rather than work around them. If we find that it does end up crashing because someone emits a progress report when we're shutting down LLDB **and** we've investigated the bug and think the functionality is important, I'm totally on board with leaking it. Does that sound like a reasonable compromise?

clayborg replied:
It is fine to start with this and switch over eventually if we run into problems. Not sure what percentage of people will listen to the category progresses vs the individual ones. So it will affect only the people using the category ones.

clayborg replied:
But wether you leak it or not it won't affect the program since it is exiting already when this gets cleaned up if it is left as is above. So up to you guys.

chelcassanova replied:
So we'll keep the current functionality (using `Initialize/Terminate`, and having the user check the optional instead of doing it in the class) if I'm understanding correctly?

clayborg replied:
Works for me as long as all clients check the optional and use it correctly. It allows for multi-thread safety so that is good. Not sure if std::optional is actually threadsafe though, but the timing required to make this fail would be pretty low percentage. If it were me I would make it bullet proof as just leak the map as when we tear down the process, any allocated memory will just go poof anyway and it will actually speed up the process teardown process as no work will need to be done to clear this map, though it should really be empty anyway.

chelcassanova replied:
Given that this patch follows a pattern established by the other subsystems, and those subsystems don't currently show any major issues in production with their initialization and termination lifecycle I think we can follow the current subsystem behaviour for this patch and discussions about the subsystems themselves can take place in a separate discussion.

chelcassanova committed:
Use call_once in InstanceImpl(), leak progress manager object

In order to prevent potential crashes when LLDB is terminated,
this commit instead creates the instance for the progress manager in a
`call_once` within its initializer.

clayborg reviewed (approved):
Looks good!

JDevlieghere commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -119,6 +120,32 @@ class Progress {
+  ProgressManager();
+  ~ProgressManager();
+
+  static void Initialize();
+  static void Terminate();
```
These are not implemented (and no longer needed)?

chelcassanova replied:
Correct, I'll remove these.

JDevlieghere reviewed (approved):
LGTM with the Initialize/Terminate removed as they're not needed anymore.

adrian-prantl commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -119,6 +120,32 @@ class Progress {
+
+  static void Initialize();
+  static void Terminate();
+  // Control the refcount of the progress report category as needed
```
///

adrian-prantl commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -119,6 +120,32 @@ class Progress {
+  void Increment(std::string category);
+  void Decrement(std::string category);
+
+  // Public accessor for the class instance
```
/// and . at the end of the sentence.

JDevlieghere replied:
Just remove this, they don't add anything and we're slowly removing them.

adrian-prantl commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -119,6 +120,32 @@ class Progress {
+  static ProgressManager &Instance();
+
+private:
+  // Manage the class instance internally using a std::optional
```
ditto

adrian-prantl commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +67,41 @@ void Progress::ReportProgress() {
+  auto pair = m_progress_category_map.insert(std::pair(title, 1));
+
+  // If pair.first is not empty after insertion it means that that
+  // category was entered for the first time and should not be incremented
```
. at the end of sentence

adrian-prantl commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +67,41 @@ void Progress::ReportProgress() {
+  if (pos == m_progress_category_map.end())
+    return;
+
+  // Remove the category from the map if the refcount reaches 0
```
same here

adrian-prantl commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +67,41 @@ void Progress::ReportProgress() {
+    return;
+
+  // Remove the category from the map if the refcount reaches 0
+  if (--pos->second == 0)
```
What guards against pos->second being 0? Should that be an assertion at least?

chelcassanova replied:
I added a check to return early if `pos->second == 0`.

JDevlieghere commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +67,41 @@ void Progress::ReportProgress() {
+  // If pair.first is not empty after insertion it means that that
+  // category was entered for the first time and should not be incremented.
+  if (!pair.second)
+    ++pair.first->second;
```
Any reason this should be a pre-increment? `pair.first` returns an iterator so without checking the C++ reference, I don't know if this is incrementing the or the value. Thinking about it some more, you can simplify this a lot:

```
std::lock_guard<std::mutex> lock(m_progress_map_mutex);
m_progress_category_map[title]++;
```

`operator[]` will return a default constructed value (i.e. 0) and then the increment will set it to one.

chelcassanova replied:
Nice, no need to check the status of the map insertion in the first place.

JDevlieghere commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +67,41 @@ void Progress::ReportProgress() {
+  std::lock_guard<std::mutex> lock(m_progress_map_mutex);
+  auto pos = m_progress_category_map.find(title);
+
+  if (pos == m_progress_category_map.end() || pos->second == 0)
```
If `pos->second` is `0`, don't we still need to remove it? How about unconditionally removing it when it's `<= 1` and otherwise decrementing it?

chelcassanova committed:
Simpify Increment() and Decrement()

Increment() now uses the operator[] which removes the need that the
insertion in the map was of a new category. Decrement() removes the
category if its refcount is <= 1, otherwise it will then decrement it.

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +67,37 @@ void Progress::ReportProgress() {
+
+ProgressManager::~ProgressManager() {}
+
+ProgressManager &ProgressManager::Instance() { return InstanceImpl(); }
```
Remove `ProgressManager &ProgressManager::InstanceImpl()` and inline the code from that function in this function.

clayborg reviewed (commented):
Just remove `static ProgressManager &InstanceImpl();` and inline the code into `static ProgressManager &Instance();` and this is good to go.

chelcassanova commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -119,6 +120,32 @@ class Progress {
+  static ProgressManager &Instance();
+
+private:
+  // Manage the class instance internally using a std::optional
```
I should've changed this comment in a more recent push that removes the reference to the `std::optional`.

chelcassanova commented on lldb/source/Core/Progress.cpp:
```diff
@@ -66,3 +67,41 @@ void Progress::ReportProgress() {
   }
 }
+
+ProgressManager &ProgressManager::InstanceImpl() {
+  static std::once_flag g_once_flag;
+  static ProgressManager *g_progress_manager = nullptr;
+  std::call_once(g_once_flag, []() {
+    // NOTE: known leak to avoid global destructor chain issues.
+    g_progress_manager = new ProgressManager();
+  });
+  return *g_progress_manager;
+}
```
I was actually going to ask if `Instance()` was still necessary given that all functionality was in `InstanceImpl()` (or if `InstanceImpl()` now needed to be public or something along those lines). If we're going to keep the implementation of the instance accessor this way then to my understanding we can inline `Instance()`.

chelcassanova committed:
Remove InstanceImpl()

Removes `InstanceImpl()` and inlines its code into `Instance()`.
